src/vm/database/structures.rs

```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
 clarity_serializable!(u128);
 clarity_serializable!(u64);
 clarity_serializable!(Contract);
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct STXBalance {
+    pub amount_unlocked: u128,
+    pub amount_locked: u128,
+    pub unlock_height: u64,
+}
+clarity_serializable!(STXBalance);
+
+#[derive(Debug)]
+pub enum STXBalanceError {
+    Overflow,
+    Underflow,
+    LockActive,
+}
+
+type Result<T> = std::result::Result<T, STXBalanceError>;
+
+impl STXBalance {
+
+    pub fn zero() -> STXBalance {
+        STXBalance {
+            amount_unlocked: 0,
+            amount_locked: 0,
+            unlock_height: 0,
+        }
+    }
+
+    pub fn get_available_balance_at_block(&self, block_height: u64) -> u128 {
+        match self.has_locked_tokens_unlockable(block_height) {
+            true => self.get_total_balance(),
+            false => self.amount_unlocked
+        }
+    }
+
+    pub fn lock_tokens(&mut self, amount_to_lock: u128, unlock_height: u64, current_height: u64) -> Result<()> {
+        let unlocked = self.unlock_available_tokens_if_any(current_height);
+        if  unlocked > 0 {
+            debug!("Consolidated after account-token-lock");
+        }
+
+        // todo(ludo): Also check that unlock_height > current_height
```

Comment:
Please address this.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/7cf92506281c3a4e4c77f54f741ecd7999d8d03f